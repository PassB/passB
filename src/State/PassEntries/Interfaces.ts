import {List, Map, OrderedMap} from 'immutable';
import {ActionName, ExtensionName} from 'State/Interfaces';
import {TypedMap} from 'State/Types/TypedMap';

export interface ActionInterface {
  extension: ExtensionName;
  action: ActionName;
}
export type Action = TypedMap<ActionInterface>;

type EntryName = string;

export interface EntryNodeInterface {
  name: EntryName;
  fullPath: string;
  actions: List<Action>;
  children: OrderedMap<string, EntryNode>;
}

export type EntryNode = TypedMap<EntryNodeInterface>;

export interface ExtensionNameArgs {
  extensionName: ExtensionName;
}

export interface EntryAction {
  fullPath: string;
  action: ActionName;
}

export interface ExtensionEntriesArgs {
  extensionName: ExtensionName;
  entries: EntryAction[];
}

export type PassEntryState = Map<ExtensionName, EntryNode>;
