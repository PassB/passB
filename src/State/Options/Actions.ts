import {actionCreatorFactory} from 'typescript-fsa';
import {ExtensionNameArgs, ExtensionOptionsArgs, StrategyNameArgs, StrategyOptionsArgs} from './Interfaces';

const actionCreator = actionCreatorFactory('OPTIONS');

export const actions = {
  setExtensionDefaultOptions:
    actionCreator<ExtensionOptionsArgs>('SET_EXTENSION_DEFAULT_OPTIONS'),
  setExtensionOptions:
    actionCreator<ExtensionOptionsArgs>('SET_EXTENSION_OPTIONS'),
  enableExtension:
    actionCreator<ExtensionNameArgs>('ENABLE_EXTENSION'),
  disableExtension:
    actionCreator<ExtensionNameArgs>('DISABLE_EXTENSION'),
  setStrategyDefaultOptions:
    actionCreator<StrategyOptionsArgs>('SET_STRATEGY_DEFAULT_OPTIONS'),
  setStrategyOptions:
    actionCreator<StrategyOptionsArgs>('SET_STRATEGY_OPTIONS'),
  setSelectedStrategy:
    actionCreator<StrategyNameArgs>('SET_SELECTED_STRATEGY'),
};
