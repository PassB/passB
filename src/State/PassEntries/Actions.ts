import {actionCreatorFactory} from 'typescript-fsa';
import {ExtensionEntriesArgs, ExtensionNameArgs} from './Interfaces';

const actionCreator = actionCreatorFactory('PASS_ENTRIES');

export const resetEntries = actionCreator<ExtensionNameArgs>('RESET_ENTRIES');
export const setEntries = actionCreator<ExtensionEntriesArgs>('SET_ENTRIES');
