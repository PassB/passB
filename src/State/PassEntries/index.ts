export {PassEntryState} from './Interfaces';
export {reducer} from './Reducers';
export * from './Actions';
export * from './Selectors';
