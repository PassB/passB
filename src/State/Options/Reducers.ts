import {Set} from 'immutable';
import {reducerWithInitialState} from 'typescript-fsa-reducers';
import * as Actions from './Actions';
import {
  ExtensionName,
  ExtensionNameArgs,
  ExtensionOptionsArgs,
  OptionsState, OptionsStateFactory,
  StrategyNameArgs,
  StrategyOptionsArgs,
} from './Interfaces';

const setExtensionDefaultOptions =
  (oldState: OptionsState, {extensionName, options}: ExtensionOptionsArgs): OptionsState =>
    oldState.updateIn(['extensionsDefaultOptions', extensionName], () => options);

const setExtensionOptions =
  (oldState: OptionsState, {extensionName, options}: ExtensionOptionsArgs): OptionsState =>
    oldState.updateIn(['extensionsOptions', extensionName], () => options);

const enableExtension = (oldState: OptionsState, {extensionName}: ExtensionNameArgs): OptionsState =>
  oldState.updateIn(
    ['enabledExtensions'],
    (enabledExtensions: Set<ExtensionName> = Set<ExtensionName>()) => enabledExtensions.add(extensionName),
  );

const disableExtension = (oldState: OptionsState, {extensionName}: ExtensionNameArgs): OptionsState =>
  oldState.updateIn(
    ['enabledExtensions'],
    (enabledExtensions: Set<ExtensionName> = Set<ExtensionName>()) => enabledExtensions.remove(extensionName),
  );

const setStrategyDefaultOptions =
  (oldState: OptionsState, {strategyType, strategyName, options}: StrategyOptionsArgs): OptionsState =>
    oldState.updateIn(['strategyDefaultOptions', strategyType, strategyName], () => options);

const setStrategyOptions =
  (oldState: OptionsState, {strategyType, strategyName, options}: StrategyOptionsArgs): OptionsState =>
    oldState.updateIn(['strategyOptions', strategyType, strategyName], () => options);

const setSelectedStrategy = (oldState: OptionsState, {strategyType, strategyName}: StrategyNameArgs): OptionsState =>
  oldState.updateIn(['selectedStrategies', strategyType], () => strategyName);

export const reducer = reducerWithInitialState(OptionsStateFactory())
  .case(Actions.setExtensionDefaultOptions, setExtensionDefaultOptions)
  .case(Actions.setExtensionOptions, setExtensionOptions)
  .case(Actions.enableExtension, enableExtension)
  .case(Actions.disableExtension, disableExtension)
  .case(Actions.setStrategyDefaultOptions, setStrategyDefaultOptions)
  .case(Actions.setStrategyOptions, setStrategyOptions)
  .case(Actions.setSelectedStrategy, setSelectedStrategy)
  .build();
