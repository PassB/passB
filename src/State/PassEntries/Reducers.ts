import {List, OrderedMap} from 'immutable';
import {reducerWithInitialState} from 'typescript-fsa-reducers';
import {createTypedMap} from 'State/Types/TypedMap';
import * as Actions from './Actions';
import {
  Action, EntryAction,
  EntryNode,
  EntryNodeInterface,
  ExtensionEntriesArgs,
  ExtensionNameArgs,
  PassEntryState,

} from './Interfaces';

export const defaultRootNodeValues: EntryNodeInterface = {
  name: '',
  fullPath: '',
  actions: List<Action>(),
  children: OrderedMap<string, EntryNode>(),
};

export const defaultRootNode: EntryNode = createTypedMap(defaultRootNodeValues);

const getEntryNode = (data: Pick<EntryNodeInterface, 'name' | 'fullPath'> & Partial<EntryNodeInterface>): EntryNode => createTypedMap({
  ...defaultRootNodeValues,
  ...data,
});

const resetEntries =
  (oldState: PassEntryState, {extensionName}: ExtensionNameArgs): PassEntryState =>
    oldState.set(extensionName, getEntryNode({name: '', fullPath: ''}));

const setEntries = (oldState: PassEntryState, {extensionName, entries}: ExtensionEntriesArgs): PassEntryState =>
  oldState.set(extensionName, entries.reduce(
    (rootNode: EntryNode, entry: EntryAction) => {
      const isFolder = entry.fullPath.endsWith('/');
      const parts = entry.fullPath.split('/');
      if (isFolder || entry.fullPath.length === 0) {
        parts.pop();
      }

      const path: string[] = [];
      let lastPart = '';
      let accumulatedPath = '';
      let part;

      do {
        if (!rootNode.getIn(path)) {
          rootNode = rootNode.setIn(path, getEntryNode({name: lastPart, fullPath: accumulatedPath}));
        }

        part = parts.shift();
        if (!part) {
          rootNode = rootNode.updateIn(
            [...path, 'actions'],
            (actions: List<Action>) => actions.push(createTypedMap({extension: extensionName, action: entry.action})),
          );
        } else {
          lastPart = (parts.length > 0 || isFolder) ? `${part}/` : part;
          path.push('children', lastPart);
          accumulatedPath = accumulatedPath + lastPart;
        }
      } while (part);

      return rootNode;
    },
    getEntryNode({name: '', fullPath: ''}),
  ));

export const reducer = reducerWithInitialState(OrderedMap() as PassEntryState)
  .case(Actions.resetEntries, resetEntries)
  .case(Actions.setEntries, setEntries)
  .build();
