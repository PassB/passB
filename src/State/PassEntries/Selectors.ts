import {List} from 'immutable';
import {createSelector} from 'reselect';
import {ExtensionName} from 'State/Interfaces';
import {getPassEntryState} from 'State/Selectors';
import {getEnabledExtensions} from '../Options/Selectors';
import {EntryNode, PassEntryState} from './Interfaces';
import {defaultRootNode} from './Reducers';

export const getEnabledRootNodes = createSelector(
  getPassEntryState,
  getEnabledExtensions,
  (state: PassEntryState, enabledExtensions: List<ExtensionName>) =>
    state.filter((rootNode: EntryNode, extensionName: ExtensionName) => enabledExtensions.includes(extensionName)),
);

export const getCombinedPassEntries = createSelector(
  getEnabledRootNodes,
  (rootNodes: PassEntryState) => defaultRootNode.mergeDeep(...Array.from(rootNodes.values())),
);
