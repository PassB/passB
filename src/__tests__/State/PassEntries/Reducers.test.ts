import * as Actions from 'State/PassEntries/Actions';
import {reducer} from 'State/PassEntries/Reducers';

describe('PassEntries Reducer', () => {
  it('has an empty map as initial state', () => {
    expect(reducer(undefined!, {type: 'SOME-ACTION'})).toMatchSnapshot();
  });

  it('resets extension "testExtension"', () => {
    expect(reducer(undefined!, Actions.resetEntries({extensionName: 'testExtension'}))).toMatchSnapshot();
  });

  describe('tree building', () => {
    it('builds an empty tree', () => {
      expect(reducer(undefined!, Actions.setEntries({
        extensionName: 'testExtension',
        entries: [],
      }))).toMatchSnapshot();
    });

    it('adds actions to root node', () => {
      expect(reducer(undefined!, Actions.setEntries({
        extensionName: 'testExtension',
        entries: [
          {fullPath: '', action: 'test'},
          {fullPath: '', action: 'test2'},
        ],
      }))).toMatchSnapshot();
    });

    describe('directories', () => {
      it('adds actions', () => {
        expect(reducer(undefined!, Actions.setEntries({
          extensionName: 'testExtension',
          entries: [
            {fullPath: 'dir/', action: 'test'},
            {fullPath: 'foo/dir/', action: 'test2'},
            {fullPath: 'foo/bar/dir/', action: 'test3'},
          ],
        }))).toMatchSnapshot();
      });

      it('creates missing intermediate directory nodes if necessary', () => {
        expect(reducer(undefined!, Actions.setEntries({
          extensionName: 'testExtension',
          entries: [
            {fullPath: 'foo/bar/dir/', action: 'test'},
          ],
        }))).toMatchSnapshot();
      });
    });

    describe('files', () => {
      it('adds actions to file nodes', () => {
        expect(reducer(undefined!, Actions.setEntries({
          extensionName: 'testExtension',
          entries: [
            {fullPath: 'file', action: 'test'},
            {fullPath: 'foo/file', action: 'test2'},
            {fullPath: 'foo/bar/file', action: 'test3'},
          ],
        }))).toMatchSnapshot();
      });

      it('creates missing intermediate directory nodes if necessary', () => {
        expect(reducer(undefined!, Actions.setEntries({
          extensionName: 'testExtension',
          entries: [
            {fullPath: 'foo/bar/file', action: 'test'},
          ],
        }))).toMatchSnapshot();
      });
    });

    it('adds actions for similarly named directories and files into different nodes', () => {
      expect(reducer(undefined!, Actions.setEntries({
        extensionName: 'testExtension',
        entries: [
          {fullPath: 'foo/bar', action: 'fileTest'},
          {fullPath: 'foo/bar/', action: 'directoryTest'},
          {fullPath: 'foo/bar/contents', action: 'fileTest2'},
        ],
      }))).toMatchSnapshot();
    });
  });
});
