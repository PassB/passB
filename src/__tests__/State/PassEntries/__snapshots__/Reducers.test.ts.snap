// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PassEntries Reducer has an empty map as initial state 1`] = `Immutable.OrderedMap {}`;

exports[`PassEntries Reducer resets extension "testExtension" 1`] = `
Immutable.OrderedMap {
  "testExtension": Immutable.Map {
    "name": "",
    "fullPath": "",
    "actions": Immutable.List [],
    "children": Immutable.OrderedMap {},
  },
}
`;

exports[`PassEntries Reducer tree building adds actions for similarly named directories and files into different nodes 1`] = `
Immutable.OrderedMap {
  "testExtension": Immutable.Map {
    "name": "",
    "fullPath": "",
    "actions": Immutable.List [],
    "children": Immutable.OrderedMap {
      "foo/": Immutable.Map {
        "name": "foo/",
        "fullPath": "foo/",
        "actions": Immutable.List [],
        "children": Immutable.OrderedMap {
          "bar": Immutable.Map {
            "name": "bar",
            "fullPath": "foo/bar",
            "actions": Immutable.List [
              Immutable.Map {
                "extension": "testExtension",
                "action": "fileTest",
              },
            ],
            "children": Immutable.OrderedMap {},
          },
          "bar/": Immutable.Map {
            "name": "bar/",
            "fullPath": "foo/bar/",
            "actions": Immutable.List [
              Immutable.Map {
                "extension": "testExtension",
                "action": "directoryTest",
              },
            ],
            "children": Immutable.OrderedMap {
              "contents": Immutable.Map {
                "name": "contents",
                "fullPath": "foo/bar/contents",
                "actions": Immutable.List [
                  Immutable.Map {
                    "extension": "testExtension",
                    "action": "fileTest2",
                  },
                ],
                "children": Immutable.OrderedMap {},
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`PassEntries Reducer tree building adds actions to root node 1`] = `
Immutable.OrderedMap {
  "testExtension": Immutable.Map {
    "name": "",
    "fullPath": "",
    "actions": Immutable.List [
      Immutable.Map {
        "extension": "testExtension",
        "action": "test",
      },
      Immutable.Map {
        "extension": "testExtension",
        "action": "test2",
      },
    ],
    "children": Immutable.OrderedMap {},
  },
}
`;

exports[`PassEntries Reducer tree building builds an empty tree 1`] = `
Immutable.OrderedMap {
  "testExtension": Immutable.Map {
    "name": "",
    "fullPath": "",
    "actions": Immutable.List [],
    "children": Immutable.OrderedMap {},
  },
}
`;

exports[`PassEntries Reducer tree building directories adds actions 1`] = `
Immutable.OrderedMap {
  "testExtension": Immutable.Map {
    "name": "",
    "fullPath": "",
    "actions": Immutable.List [],
    "children": Immutable.OrderedMap {
      "dir/": Immutable.Map {
        "name": "dir/",
        "fullPath": "dir/",
        "actions": Immutable.List [
          Immutable.Map {
            "extension": "testExtension",
            "action": "test",
          },
        ],
        "children": Immutable.OrderedMap {},
      },
      "foo/": Immutable.Map {
        "name": "foo/",
        "fullPath": "foo/",
        "actions": Immutable.List [],
        "children": Immutable.OrderedMap {
          "dir/": Immutable.Map {
            "name": "dir/",
            "fullPath": "foo/dir/",
            "actions": Immutable.List [
              Immutable.Map {
                "extension": "testExtension",
                "action": "test2",
              },
            ],
            "children": Immutable.OrderedMap {},
          },
          "bar/": Immutable.Map {
            "name": "bar/",
            "fullPath": "foo/bar/",
            "actions": Immutable.List [],
            "children": Immutable.OrderedMap {
              "dir/": Immutable.Map {
                "name": "dir/",
                "fullPath": "foo/bar/dir/",
                "actions": Immutable.List [
                  Immutable.Map {
                    "extension": "testExtension",
                    "action": "test3",
                  },
                ],
                "children": Immutable.OrderedMap {},
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`PassEntries Reducer tree building directories creates missing intermediate directory nodes if necessary 1`] = `
Immutable.OrderedMap {
  "testExtension": Immutable.Map {
    "name": "",
    "fullPath": "",
    "actions": Immutable.List [],
    "children": Immutable.OrderedMap {
      "foo/": Immutable.Map {
        "name": "foo/",
        "fullPath": "foo/",
        "actions": Immutable.List [],
        "children": Immutable.OrderedMap {
          "bar/": Immutable.Map {
            "name": "bar/",
            "fullPath": "foo/bar/",
            "actions": Immutable.List [],
            "children": Immutable.OrderedMap {
              "dir/": Immutable.Map {
                "name": "dir/",
                "fullPath": "foo/bar/dir/",
                "actions": Immutable.List [
                  Immutable.Map {
                    "extension": "testExtension",
                    "action": "test",
                  },
                ],
                "children": Immutable.OrderedMap {},
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`PassEntries Reducer tree building files adds actions to file nodes 1`] = `
Immutable.OrderedMap {
  "testExtension": Immutable.Map {
    "name": "",
    "fullPath": "",
    "actions": Immutable.List [],
    "children": Immutable.OrderedMap {
      "file": Immutable.Map {
        "name": "file",
        "fullPath": "file",
        "actions": Immutable.List [
          Immutable.Map {
            "extension": "testExtension",
            "action": "test",
          },
        ],
        "children": Immutable.OrderedMap {},
      },
      "foo/": Immutable.Map {
        "name": "foo/",
        "fullPath": "foo/",
        "actions": Immutable.List [],
        "children": Immutable.OrderedMap {
          "file": Immutable.Map {
            "name": "file",
            "fullPath": "foo/file",
            "actions": Immutable.List [
              Immutable.Map {
                "extension": "testExtension",
                "action": "test2",
              },
            ],
            "children": Immutable.OrderedMap {},
          },
          "bar/": Immutable.Map {
            "name": "bar/",
            "fullPath": "foo/bar/",
            "actions": Immutable.List [],
            "children": Immutable.OrderedMap {
              "file": Immutable.Map {
                "name": "file",
                "fullPath": "foo/bar/file",
                "actions": Immutable.List [
                  Immutable.Map {
                    "extension": "testExtension",
                    "action": "test3",
                  },
                ],
                "children": Immutable.OrderedMap {},
              },
            },
          },
        },
      },
    },
  },
}
`;

exports[`PassEntries Reducer tree building files creates missing intermediate directory nodes if necessary 1`] = `
Immutable.OrderedMap {
  "testExtension": Immutable.Map {
    "name": "",
    "fullPath": "",
    "actions": Immutable.List [],
    "children": Immutable.OrderedMap {
      "foo/": Immutable.Map {
        "name": "foo/",
        "fullPath": "foo/",
        "actions": Immutable.List [],
        "children": Immutable.OrderedMap {
          "bar/": Immutable.Map {
            "name": "bar/",
            "fullPath": "foo/bar/",
            "actions": Immutable.List [],
            "children": Immutable.OrderedMap {
              "file": Immutable.Map {
                "name": "file",
                "fullPath": "foo/bar/file",
                "actions": Immutable.List [
                  Immutable.Map {
                    "extension": "testExtension",
                    "action": "test",
                  },
                ],
                "children": Immutable.OrderedMap {},
              },
            },
          },
        },
      },
    },
  },
}
`;
